docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
 
 .. _sparse-docs:
 
 torch.sparse
 ============
 
-.. warning::
+Torch supports sparse tensors which provide memory-efficient storage
+and processing algorithms of tensors for which the majority of
+elements have the same value (fill value). This allows saving storage
+and processing resources when the size and sparsity levels of tensors
+are high.  Otherwise, for small-sized or low-sparsity tensors using
+strided tensors is likely the most efficient approach.
 
-    This API is in beta and may change in the near future.
+.. _sparse-fill-value-docs:
 
-Torch supports sparse tensors in COO(rdinate) format, which can
-efficiently store and process tensors for which the majority of elements
-are zeros.
+Fill value
+++++++++++
 
-A sparse tensor is represented as a pair of dense tensors: a tensor
-of values and a 2D tensor of indices.  A sparse tensor can be constructed
-by providing these two tensors, as well as the size of the sparse tensor
-(which cannot be inferred from these tensors!)  Suppose we want to define
-a sparse tensor with the entry 3 at location (0, 2), entry 4 at
-location (1, 0), and entry 5 at location (1, 2).  We would then write:
+In Torch, the fill value is not specified explicitly but is implicitly
+interpreted by sparse algorithms. The implicit interpretation may vary
+in-between algorithms of different fields and applications. The most
+commonly assumed fill value is zero and it is assumed for all tools
+that are available via :mod:`torch` namespace. When some sparse tensor
+functionality assumes a non-zero or indefinite fill value, the
+corresponding tools are provided via :mod:`torch.sparse`
+namespace. Please see the documentation strings of the corresponding
+functions for a particular definition of the fill value.
 
-    >>> i = torch.LongTensor([[0, 1, 1],
-                              [2, 0, 2]])
-    >>> v = torch.FloatTensor([3, 4, 5])
-    >>> torch.sparse.FloatTensor(i, v, torch.Size([2,3])).to_dense()
-     0  0  3
-     4  0  5
-    [torch.FloatTensor of size 2x3]
+.. _sparse-hybrid-docs:
 
-Note that the input to LongTensor is NOT a list of index tuples.  If you want
+Hybrid tensors
+++++++++++++++
+
+Torch supports hybrid sparse tensors with dimensions split into sparse
+dimensions and dense dimensions. Dense dimensions always follow sparse
+dimensions, that is, mixing of dense and sparse dimensions is not
+supported. One can interpret hybrid tensors as sparse tensors with
+elements being (contiguous) strided tensors. We use (M +
+K)-dimensional tensor to denote a N-dimensional hybrid sparse tensor
+where M and K are the number of sparse and dense dimensions,
+respectively, such that M + K == N holds.
+
+See :ref:`Hybrid COO tensors <sparse-hybrid-coo-docs>`.
+
+Sparse tensor formats
++++++++++++++++++++++
+
+Torch supported sparse storage formats are:
+
+- COO(rdinate) format where tensor elements are stored using the lists
+  of indices and the corresponding values. Unspecified elements have
```

Comment:
This first sentence needs to change because there are no "(the) lists of indices" or "corresponding values" introduced to the reader. Maybe the structure can be:

- Sparse tensors in PyTorch use the coordinate list (COO) format
- In this format, an n-dimensional sparse tensor is represented using two lists of equal length. The first list is of n-element vectors representing "indices" that specify the position of the "values" in the second list.
- These lists are represented using two strided tensors (one of size n x values, the other of size 1 x values)
- Unspecified values are assumed to all have the same value, known as the "fill value"
- For example, ...
- When a tensor is large but many elements have the same value, this format can use less memory than a strided representation of the same tensor (maybe put the equation, assuming a coalesced tensor?)

Should the coalesced discussion and an example occur here?

Reply:
While "Sparse tensors in PyTorch use the coordinate list (COO) format" is the current state, I think this state needs to be changed as soon as possible because the COO format is known to be a simple but also one of the least efficient way to process sparse arrays in Linear Algebra operations. For instance, the matmul operation using GCS format is more than 100x more efficient than when using COO format, see the preliminary results [here](https://github.com/pytorch/pytorch/pull/44190#issuecomment-694771306) and [here](https://github.com/pytorch/pytorch/pull/44190#issuecomment-701924345).

That said, @mruberry, thanks for outlining the possible structure for the COO format. I would only add that processing of arrays in sparse formats efficiently is an important aspect in addition to using less memory for storing sparse arrays.

Reply:
That sounds good. I agree that more formats are coming... a sentence to that effect could be added? 

As long as we're careful about the performance claims then talking about performance seems fine.